guides/micronaut-openapi-generator-server/micronaut-openapi-generator-server.adoc

```diff
@@ -0,0 +1,550 @@
+<2> We specify that a body is required and what are the supported content-types for it. (in this case only `application/json`, but multiple can be allowed).
+<3> We write that `BookInfo` object is required to be in the request body. We reference the same `BookInfo` schema that we will define next.
+
+NOTE: To read more about body definitions, see the
+link:https://swagger.io/docs/specification/describing-request-body/["Describing Request Body" OpenAPI guide].
+
+=== Defining Schemas
+
+Schemas are required whenever a parameter, request body or a response body we want to describe needs to be an object. In that case we add a schema that defines all the properties of the object. You can find out about the format for schemas in the link:https://oai.github.io/Documentation/specification-content.html["Content of Message Bodies" OpenAPI Guide].
+
+We will add schemas to our definition file:
+
+resource:library-definition.yml[tag=components]
+
+<1> We define the `BookInfo` schema inside then `components/schemas` section. From this schema a Java class will be generated with the same `BookInfo` class name.
+<2> We define all the properties of `BookInfo`, including required validation on them (In this case, it is a minimal length requirement on one string and a regex pattern on another). An abbreviated form is used for some YAML lists and dictionaries to reduce the number of rows and simplify readability.
+<3> We reference another schema to be used as a property.
+<4> We define `BookAvailability` schema to be an enum with three available values. A Java `BookAvailability` class will be generated with given enum values based on our definition.
+
+As you can see, schemas can be defined as enums when they can only be assigned a finite number of values. Also, you can reference other schemas as properties of a schema.
+
+NOTE: You can read more about writing schemas in the
+link:https://swagger.io/docs/specification/data-models/["Data Models (Schemas)" OpenAPI guide].
+
+Save the file and proceed to the next part of the guide.
+//Altogether it should look like this: link:@sourceDir@/library-definition.yml[library-definition.yml].
+
+== Generating Server API From The OpenAPI Definition
+
+Now we will generate server API files from our definition. The generated server code will be in Java and will use the Micronaut features for client-server communication.
+
+:exclude-for-build:maven
+
+Open your `build.gradle` file an apply the `micronaut-openapi` plugin:
+
+
+[source, groovy]
+.build.gradle
+----
+plugins {
+  id 'io.micronaut.openapi' version '...'
+  ....
+}
+----
+
+And configure your build to generate a client:
+
+[source, groovy]
+.build.gradle
+----
+micronaut {
+  ...
+  openapi {
+    server(file("src/main/resources/library-definition.yml")) { // <1>
+      apiPackageName = "example.micronaut.api"
+      modelPackageName = "example.micronaut.model"
+      invokerPackage = "example.micronaut.invoker"
```

Comment:
I'm confused, where does the `annotation` package come from?

Reply:
It comes from inside the invoker package. The invoker package has all the invocation and utility classes (had the `Application.java` previously and can have some security configuration for the client).
The `HardNullable` is an inherited nullable generated in `{invokerPackage}.annotation`.

Reply:
I understand but you said "use `micronaut.example` for the example package", not `micronaut.example.annotation`.

Reply:
The `HardNullable` is put in a separate subpackage to make the project better structured in case there are other files in the invoker package (in this case there are none, so it looks strange).

Reply:
Oh. That's quite surprising behavior. But I think it won't work here, since we're generating classes in an isolated directory in `build/...` so there won't be other types in that package.

Reply:
Currently, it works fine:
![image](https://github.com/micronaut-projects/micronaut-guides/assets/80816836/43bfd632-7317-49af-b15d-c3c597383c55)

As I understand this PR is not used yet: https://github.com/micronaut-projects/micronaut-gradle-plugin/pull/758. But it seems like [this line](https://github.com/micronaut-projects/micronaut-gradle-plugin/blob/204702be5d866419e3b577bb8990a9b41e0cd634/openapi-plugin/src/main/java/io/micronaut/gradle/openapi/DefaultOpenApiExtension.java#L80) adds supporting files to the api output, so the `HardNullable` should be generated then.

Reply:
the PR is used. I am extremely confused by the behavior. You are saying, basically, that if the invoker package is set to `example.micronaut`, then the annotation should be generated under `example.micronaut.annotation` _because_ there are other files in that directory. That is quite surprising behavior to me, as I would expect the generator to honor what I tell it to use. I also understand that "not empty" only works because by accident, the `api` and `model` types are created before the supporting files. This wouldn't be the case, for example, if we had 2 tasks, one to generate the `api`, the other to generate the supporting files. So I would rather use a `example.micronaut.invoker` package, so that the behavior is consistent.

Reply:
No, it is just generated in the `example.micronaut.annotation` directory solely based on the fact that the invoker package is `example.micronaut` and independently of if there are other files in the directory.

Reply:
Then I don't understand. Why doesn't it generate something in `example.micronaut.invoker.annotation` if I set the package to `example.micronaut.invoker`? This seems to contradict what I'm effectively seeing.

Reply:
Sorry for the confusion... This is what it will generate if we set the invoker package to `example.micronaut`:
> The invoker package is different for gradle and maven. Probably best to set it to simply example.micronaut common package.

Reply:
Ok thanks, I fixed the (non intended) difference in the Maven example, and used `example.micronaut` in both cases for the invoker package.

Reply:
Thank you!